main/test/ca/uwaterloo/flix/language/phase/TestTyper.scala

```diff
@@ -1068,4 +1068,30 @@ class TestTyper extends FunSuite with TestUtils {
     expectError[TypeError.MismatchedBools](result)
   }
 
+  test("Test.IllegalMain.01") {
+    val input =
+      """
+        |def main(blah: Array[String]): Int32 = ???
+        |""".stripMargin
+    val result = compile(input, DefaultOptions)
+    expectError[TypeError.IllegalMain](result)
+  }
+
+  test("Test.IllegalMain.02") {
+    val input =
+      """
+        |def main(blah: Array[Char]): Int32 & Impure = ???
+        |""".stripMargin
+    val result = compile(input, DefaultOptions)
+    expectError[TypeError.IllegalMain](result)
+  }
+
+  test("Test.IllegalMain.03") {
+    val input =
+      """
+        |def main(blah: Array[String]): Int64 & Impure = ???
+        |""".stripMargin
+    val result = compile(input, DefaultOptions)
+    expectError[TypeError.IllegalMain](result)
+  }
```

Comment:
Two more cases:

- Polymorphism in the type of the array/return type.
- TC constraints.
:)

Reply:
Sorry to be pedantic, but can you also check that polymorphism is rejected in the *return type*. (I have had plenty of such issues)

Reply:
Added.